dsp/modules/llamaindex.py

```diff
@@ -0,0 +1,335 @@
+"""
+this code implements a wrapper around the llama_index library to emulate a dspy llm
+
+this allows the llama_index library to be used in the dspy framework since dspy has limited support for LLMs
+
+This code is a slightly modified copy of dspy/dsp/modules/azure_openai.py
+
+The way this works is simply by creating a dummy openai client that wraps around any llama_index LLM object and implements .complete and .chat
+
+tested with python 3.12
+
+dspy==0.1.4
+dspy-ai==2.4.9
+llama-index==0.10.35
+llama-index-llms-openai==0.1.18
+
+"""
+
+import json
+import logging
+from typing import Any, Literal
+
+from easydict import EasyDict
+from llama_index.core.base.llms.types import ChatMessage
+from llama_index.core.llms import LLM
+
+
+def LlamaIndexOpenAIClientWrapper(llm: LLM):
```

Comment:
To clarify, does this support compatibility with all Llamaindex-LLM integrations? It seems to me like the definition here is only for OpenAI clients, which would leave this change redundant to just using DSPy.OpenAI. If it is not just isolated to OpenAI, lets refactor this function to LlamaIndexLLMWrapper

Reply:
This is supposed to support all llamaindex llms. However the way i use this is by wrapping the llamaindex llm to look like an `openai` client object, then i pass this to a dspy azure openai llm.
Llamaindex llm.chat and llm.complete are used for the actual calls